File: evd_ros_backend/evd_ros_core/src/evd_script/program/primitives/gripper.py
from ..primitive import Primitive


class Gripper(Primitive):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls):
        return 'gripper.'

    @classmethod
    def full_type_string(cls):
        return Primitive.full_type_string() + cls.type_string()

    def __init__(self, position=0, effort=0, speed=0, thing_uuid=None, type='',
                 name='', uuid=None, parent=None, append_type=True):

        self._thing_uuid = None
        self._position = None
        self._effort = None
        self._speed = None

        super(Gripper,self).__init__(
            type='gripper.'+type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self.thing_uuid = thing_uuid
        self.position = position
        self.effort = effort
        self.speed = speed

    def to_dct(self):
        msg = super(Gripper,self).to_dct()
        msg.update({
            'thing_uuid': self.thing_uuid,
            'position': self.position,
            'effort': self.effort,
            'speed': self.speed
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            name=dct['name'],
            uuid=dct['uuid'],
            type=dct['type'],
            append_type=False,
            position=dct['position'],
            effort=dct['effort'],
            speed=dct['speed'],
            thing_uuid=dct['thing_uuid'])

    '''
    Data accessor/modifier methods
    '''

    @property
    def position(self):
        return self._position

    @position.setter
    def position(self, value):
        if self._position != value:
            self._position = value
            self.updated_attribute('position','set')

    @property
    def effort(self):
        return self._effort

    @effort.setter
    def effort(self, value):
        if self._effort != value:
            self._effort = value
            self.updated_attribute('effort','set')

    @property
    def speed(self):
        return self._speed

    @speed.setter
    def speed(self, value):
        if self._speed != value:
            self._speed = value
            self.updated_attribute('speed','set')

    @property
    def thing_uuid(self):
        return self._thing_uuid

    @thing_uuid.setter
    def thing_uuid(self, value):
        if self._thing_uuid != value:
            self._thing_uuid = value
            self.updated_attribute('thing_uuid','set')

    def set(self, dct):
        position = dct.get('position', None)
        if position != None:
            self.position = position

        effort = dct.get('effort', None)
        if effort != None:
            self.effort = effort

        speed = dct.get('speed', None)
        if speed != None:
            self.speed = speed

        if 'thing_uuid' in dct.keys():
            self.thing_uuid = dct['thing_uuid']

        super(Gripper,self).set(dct)

    '''
    Update Methods
    '''

    def deep_update(self):

        super(Gripper,self).deep_update()

        self.updated_attribute('thing_uuid','update')
        self.updated_attribute('position','update')
        self.updated_attribute('effort','update')
        self.updated_attribute('speed','update')

    def shallow_update(self):
        super(Gripper,self).shallow_update()

        self.updated_attribute('thing_uuid','update')
        self.updated_attribute('position','update')
        self.updated_attribute('effort','update')
        self.updated_attribute('speed','update')

    '''
    Execution methods
    '''

    def symbolic_execution(self, hooks):
        pass

    def realtime_execution(self, hooks):
        pass

File: evd_authoring_app/src/model/evdScript/program/primitives/gripper.js
import { Primitive } from '../primitive';
import Blockly from 'blockly';

export class Gripper extends Primitive {

    /**
     * data structure methods
     */

    static typeString() {
        return 'gripper.';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + Gripper.typeString();
    }

    constructor(thingUuid=null, position=0, effort=0, speed=0, type='', name='', 
                uuid=null, parent=null, appendType=true) 
    {
        super(
            (appendType) ? 'gripper.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );

        this._thingUuid = null;
        this._position = null;
        this._effort = null;
        this._speed = null;

        this.thingUuid = thingUuid;
        this.position = position;
        this.effort = effort;
        this.speed = speed;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            thing_uuid: this.thingUuid,
            position: this.position,
            effort: this.effort,
            speed: this.speed
        };
        return msg;
    }

    static fromDict(dct) {
        return new Gripper(
            dct.thing_uuid,
            dct.position,
            dct.effort,
            dct.speed,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    static BlocklyToolbox() {
        return { type: 'gripper' };
    }

    static BlocklyBlock() {
        return { key: 'gripper', data: {
            init: function() {
                this.appendDummyInput()
                    .appendField("Gripper");
                this.appendValueInput("thing")
                    .setCheck("thing")
                    .appendField("Thing");
                this.appendDummyInput()
                    .appendField("Position")
                    .appendField(new Blockly.FieldNumber(50, 0, 100), "position")
                    .appendField("%");
                this.appendDummyInput()
                    .appendField("Effort")
                    .appendField(new Blockly.FieldNumber(100, 0, 100), "effort")
                    .appendField("%");
                this.appendDummyInput()
                    .appendField("Speed")
                    .appendField(new Blockly.FieldNumber(100, 0, 100), "speed")
                    .appendField("%");
                this.setInputsInline(false);
                this.setPreviousStatement(true, null);
                this.setNextStatement(true, null);
                this.setColour(120);
                this.setTooltip("gripper");
                this.setHelpUrl("gripper");
            }
        }};
    }

     /**
      * data accessor / modifier methods
      */
    
    get position() {
        return this._position;
    }

    set position(value) {
        if (this._position !== value) {
            this._position = value;
            this.updatedAttribute('position','set');
        }
    }

    get effort() {
        return this._effort;
    }

    set effort(value) {
        if (this._effort !== value) {
            this._effort = value;
            this.updatedAttribute('effort','set');
        }
    }

    get speed() {
        return this._speed;
    }

    set speed(value) {
        if (this._speed !== value) {
            this._speed = value;
            this.updatedAttribute('speed','set');
        }
    }

    get thingUuid() {
        return this._thingUuid;
    }
    
    set thingUuid(value) {
        if (this._thingUuid !== value) {
            this._thingUuid = value;
            this.updatedAttribute('thing_uuid','set');
        }
    }

    set(dct) {

        if ('position' in dct) {
            this.position = dct.position;
        }

        if ('effort' in dct) {
            this.effort = dct.effort;
        }

        if ('speed' in dct) {
            this.speed = dct.speed;
        }

        if ('thing_uuid' in dct) {
            this.thingUuid = dct.thing_uuid;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('thing_uuid','update');
        this.updatedAttribute('position','update');
        this.updatedAttribute('effort','update');
        this.updatedAttribute('speed','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('thing_uuid','update');
        this.updatedAttribute('position','update');
        this.updatedAttribute('effort','update');
        this.updatedAttribute('speed','update');
    }
}